osext_openbsd.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2012 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package osext
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"unsafe"
+)
+
+var initCwd, initCwdErr = os.Getwd()
+
+func executable() (string, error) {
+	var mib [4]int32
+	mib = [4]int32{1 /* CTL_KERN */, 55 /* KERN_PROC_ARGS */, int32(os.Getpid()), 1 /* KERN_PROC_ARGV */}
+
+	n := uintptr(0)
+	// get length
+	_, _, errNum := syscall.Syscall6(syscall.SYS___SYSCTL, uintptr(unsafe.Pointer(&mib[0])), 4, 0, uintptr(unsafe.Pointer(&n)), 0, 0)
+	if errNum != 0 {
+		return "", errNum
+	}
+	if n == 0 { // shouldn't happen
+		return "", nil
+	}
+	buf := make([]byte, n)
+	_, _, errNum = syscall.Syscall6(syscall.SYS___SYSCTL, uintptr(unsafe.Pointer(&mib[0])), 4, uintptr(unsafe.Pointer(&buf[0])), uintptr(unsafe.Pointer(&n)), 0, 0)
+	if errNum != 0 {
+		return "", errNum
+	}
+	if n == 0 { // shouldn't happen
+		return "", nil
+	}
+
+	// buf now contains **argv, with pointers to each of the C-style NULL terminated arguments.
+	var args []string
+	argv := uintptr(unsafe.Pointer(&buf[0]))
+	for {
+		argp := *(**[1<<20]byte)(unsafe.Pointer(argv))
```

Comment:
Dealing with sth that return C strings I have to slice off the trailing 0 byte.
The data returned from the KERN_PROC_ARGV sysctl is not a single string - it
is an actual char **argv so there's a series of pointers for each of the
argv values (the leading values would actually be pointers) so I have to
extract stuffs accordingly and convert it into the appropriate form.